src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfoOfT.cs

```diff
@@ -155,60 +215,76 @@ internal void InitializeForSourceGen(JsonSerializerOptions options, JsonProperty
             {
                 IsIgnored = true;
-                Debug.Assert(!ShouldSerialize);
-                Debug.Assert(!ShouldDeserialize);
+                Debug.Assert(!CanSerialize);
+                Debug.Assert(!CanDeserialize);
             }
             else
             {
                 Get = propertyInfo.Getter!;
                 Set = propertyInfo.Setter;
-                HasGetter = Get != null;
-                HasSetter = Set != null;
                 JsonTypeInfo = propertyTypeInfo;
                 DeclaringType = declaringType;
                 IgnoreCondition = propertyInfo.IgnoreCondition;
                 MemberType = propertyInfo.IsProperty ? MemberTypes.Property : MemberTypes.Field;
-                ConverterStrategy = Converter!.ConverterStrategy;
+                ConverterStrategy = typedEffectiveConverter.ConverterStrategy;
                 NumberHandling = propertyInfo.NumberHandling;
             }
         }
 
-        internal override void Configure()
+        internal override void Configure(JsonTypeInfo typeInfo)
         {
-            base.Configure();
+            base.Configure(typeInfo);
 
             if (!IsForTypeInfo && !IsIgnored)
             {
-                _converterIsExternalAndPolymorphic = !ConverterBase.IsInternalConverter && PropertyType != ConverterBase.TypeToConvert;
+                _converterIsExternalAndPolymorphic = !EffectiveConverter.IsInternalConverter && PropertyType != EffectiveConverter.TypeToConvert;
                 _propertyTypeEqualsTypeToConvert = typeof(T) == PropertyType;
             }
         }
 
-        internal override JsonConverter ConverterBase
+        internal override void SetEffectiveConverter()
+        {
+            JsonConverter? customConverter = CustomConverter;
+            if (customConverter != null)
+            {
+                customConverter = Options.ExpandFactoryConverter(customConverter, PropertyType);
+                JsonSerializerOptions.CheckConverterNullabilityIsSameAsPropertyType(customConverter, PropertyType);
+            }
+
+            JsonConverter effectiveConverter = customConverter ?? NonCustomConverter ?? Options.GetConverterForType(PropertyType);
+            if (effectiveConverter.TypeToConvert == PropertyType)
+            {
+                EffectiveConverter = effectiveConverter;
+            }
+            else
+            {
+                EffectiveConverter = effectiveConverter.CreateCastingConverter<T>();
+            }
+        }
+
+        internal override JsonConverter EffectiveConverter
         {
             get
             {
-                return Converter;
+                return TypedEffectiveConverter;
             }
             set
             {
-                Debug.Assert(value is JsonConverter<T>);
-                Converter = (JsonConverter<T>)value;
+                TypedEffectiveConverter = (JsonConverter<T>)value;
```

Comment:
Should this also set `EffectiveConverter`? You might consider using a similar strategy like the delegate setters, ensuring that each set operation keeps typed and untyped fields in sync.

Reply:
TypedEffectiveConverter is a backing field/property for EffectiveConverter